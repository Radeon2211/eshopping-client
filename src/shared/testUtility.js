// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
