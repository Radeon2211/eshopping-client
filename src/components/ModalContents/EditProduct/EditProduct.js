import { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useSelector, useDispatch } from 'react-redux';
import * as actions from '../../../store/actions/indexActions';
import Form from '../../UI/Form/Form';
import Input from '../../UI/Input/Input';
import UploadPhoto from '../../UploadPhoto/UploadPhoto';
import SideBySide from '../../UI/SideBySide';
import { inputKinds, productRules } from '../../../shared/constants';
import { getChangedValues } from '../../../shared/utility/utility';
import { ProductCondition } from '../../../shared/types/enums';

const validationSchema = Yup.object(productRules);

export default function EditProduct() {
  const productDetails = useSelector((state) => state.product.productDetails);

  const dispatch = useDispatch();
  const onEditProduct = useCallback(
    (data, productId) => dispatch(actions.editProduct(data, productId)),
    [dispatch],
  );

  const initialValues = {
    name: productDetails?.name,
    price: productDetails?.price,
    quantity: productDetails?.quantity,
    condition: productDetails?.condition,
    description: productDetails?.description,
    photo: null,
  };

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(data) => {
        const changedValues = getChangedValues(data, initialValues);
        onEditProduct(changedValues, productDetails?._id);
      }}
    >
      {({ errors, touched, dirty, isValid, setFieldTouched, setFieldValue, values }) => (
        <Form btnText="Edit" headingText="Edit a product" isValid={dirty && isValid} cancellable>
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'text',
              name: 'name',
              id: 'name',
              placeholder: 'Name of the product (up to 150 characters)',
              autoComplete: 'off',
              autoFocus: true,
              onInput: setFieldTouched.bind(this, 'name', true, true),
              'data-testid': 'EditProduct-name',
            }}
            label="Name"
            isValid={!errors.name}
            isTouched={touched.name}
          />
          <SideBySide $proportion="1/1">
            <Input
              kind={inputKinds.INPUT}
              config={{
                type: 'number',
                name: 'price',
                id: 'price',
                placeholder: 'Price of the product',
                autoComplete: 'off',
                min: 0.01,
                max: 1000000,
                step: '.01',
                onInput: setFieldTouched.bind(this, 'price', true, true),
                'data-testid': 'EditProduct-price',
              }}
              label="Price ($)"
              isValid={!errors.price}
              isTouched={touched.price}
            />
            <Input
              kind={inputKinds.INPUT}
              config={{
                type: 'number',
                name: 'quantity',
                id: 'quantity',
                placeholder: 'Quantity you want to sell',
                min: 1,
                max: 1000000,
                onInput: setFieldTouched.bind(this, 'quantity', true, true),
                'data-testid': 'EditProduct-quantity',
              }}
              label="Quantity"
              isValid={!errors.quantity}
              isTouched={touched.quantity}
            />
          </SideBySide>
          <Input
            kind={inputKinds.RADIO}
            config={{
              name: 'condition',
              value: ProductCondition.NOT_APPLICABLE,
              options: [
                {
                  value: ProductCondition.NEW,
                  id: ProductCondition.NEW,
                  checked: values.condition === ProductCondition.NEW,
                  label: ProductCondition.NEW,
                  'data-testid': 'EditProduct-condition-new',
                },
                {
                  value: ProductCondition.USED,
                  id: ProductCondition.USED,
                  checked: values.condition === ProductCondition.USED,
                  label: ProductCondition.USED,
                  'data-testid': 'EditProduct-condition-used',
                },
                {
                  value: ProductCondition.NOT_APPLICABLE,
                  id: ProductCondition.NOT_APPLICABLE,
                  checked: values.condition === ProductCondition.NOT_APPLICABLE,
                  label: 'not applicable',
                  'data-testid': 'EditProduct-condition-not-applicable',
                },
              ],
            }}
            label="Condition"
          />
          <Input
            kind={inputKinds.TEXTAREA}
            config={{
              name: 'description',
              id: 'description',
              placeholder: 'Describe your product (up to 800 characters)',
              onInput: setFieldTouched.bind(this, 'description', true, true),
              maxRows: 6,
              'data-testid': 'EditProduct-description',
            }}
            label="Description"
            isValid={!errors.description}
            isTouched={touched.description}
          />
          <UploadPhoto
            setFieldValue={setFieldValue}
            hasCurrentPhoto={Boolean(productDetails.photo)}
          />
        </Form>
      )}
    </Formik>
  );
}
