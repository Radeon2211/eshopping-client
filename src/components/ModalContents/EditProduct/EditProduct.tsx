import { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import * as actions from '../../../store/actions/indexActions';
import Form from '../../UI/Form/Form';
import Input from '../../UI/Input/Input';
import UploadPhoto from '../../UploadPhoto/UploadPhoto';
import SideBySide from '../../UI/SideBySide';
import { productRules } from '../../../shared/constants';
import { getChangedValues } from '../../../shared/utility/utility';
import { ProductCondition } from '../../../shared/types/enums';
import { InputKind } from '../../../shared/types/types';
import { useTypedDispatch, useTypedSelector } from '../../../store/reducers/rootReducer';
import { AddProductForm } from '../../../shared/types/forms';

const validationSchema = Yup.object(productRules);

export default function EditProduct() {
  const productDetails = useTypedSelector((state) => state.product.productDetails);

  if (!productDetails) return null;

  const dispatch = useTypedDispatch();
  const onEditProduct = useCallback(
    (data: Partial<AddProductForm>, productId: string) =>
      dispatch(actions.editProduct(data, productId)),
    [dispatch],
  );

  const initialValues = {
    name: productDetails.name,
    price: productDetails.price,
    quantity: productDetails.quantity,
    condition: productDetails.condition,
    description: productDetails.description,
    photo: null,
  };

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(data: AddProductForm) => {
        const changedValues = getChangedValues(data, initialValues);
        onEditProduct(changedValues, productDetails?._id);
      }}
    >
      {({ errors, touched, dirty, isValid, setFieldTouched, setFieldValue, values }) => (
        <Form btnText="Edit" headingText="Edit a product" isValid={dirty && isValid} cancellable>
          <Input
            kind={InputKind.INPUT}
            config={{
              type: 'text',
              name: 'name',
              id: 'name',
              placeholder: 'Name of the product (up to 150 characters)',
              autoComplete: 'off',
              autoFocus: true,
              onInput: setFieldTouched.bind(EditProduct, 'name', true, true),
              'data-testid': 'EditProduct-name',
            }}
            label="Name"
            isValid={!errors.name}
            isTouched={touched.name}
          />
          <SideBySide $proportion="1/1">
            <Input
              kind={InputKind.INPUT}
              config={{
                type: 'number',
                name: 'price',
                id: 'price',
                placeholder: 'Price of the product',
                autoComplete: 'off',
                min: 0.01,
                max: 1000000,
                step: '.01',
                onInput: setFieldTouched.bind(EditProduct, 'price', true, true),
                'data-testid': 'EditProduct-price',
              }}
              label="Price ($)"
              isValid={!errors.price}
              isTouched={touched.price}
            />
            <Input
              kind={InputKind.INPUT}
              config={{
                type: 'number',
                name: 'quantity',
                id: 'quantity',
                placeholder: 'Quantity you want to sell',
                min: 1,
                max: 1000000,
                onInput: setFieldTouched.bind(EditProduct, 'quantity', true, true),
                'data-testid': 'EditProduct-quantity',
              }}
              label="Quantity"
              isValid={!errors.quantity}
              isTouched={touched.quantity}
            />
          </SideBySide>
          <Input
            kind={InputKind.RADIO}
            config={{
              name: 'condition',
              options: [
                {
                  value: ProductCondition.NEW,
                  id: ProductCondition.NEW,
                  checked: values.condition === ProductCondition.NEW,
                  label: ProductCondition.NEW,
                  'data-testid': 'EditProduct-condition-new',
                },
                {
                  value: ProductCondition.USED,
                  id: ProductCondition.USED,
                  checked: values.condition === ProductCondition.USED,
                  label: ProductCondition.USED,
                  'data-testid': 'EditProduct-condition-used',
                },
                {
                  value: ProductCondition.NOT_APPLICABLE,
                  id: ProductCondition.NOT_APPLICABLE,
                  checked: values.condition === ProductCondition.NOT_APPLICABLE,
                  label: 'not applicable',
                  'data-testid': 'EditProduct-condition-not-applicable',
                },
              ],
            }}
            label="Condition"
          />
          <Input
            kind={InputKind.TEXTAREA}
            config={{
              name: 'description',
              id: 'description',
              placeholder: 'Describe your product (up to 800 characters)',
              onInput: setFieldTouched.bind(EditProduct, 'description', true, true),
              maxRows: 6,
              'data-testid': 'EditProduct-description',
            }}
            label="Description"
            isValid={!errors.description}
            isTouched={touched.description}
          />
          <UploadPhoto
            setFieldValue={setFieldValue}
            hasCurrentPhoto={Boolean(productDetails.photo)}
          />
        </Form>
      )}
    </Formik>
  );
}
