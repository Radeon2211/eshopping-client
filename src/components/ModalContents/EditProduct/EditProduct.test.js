import React from 'react';
import { render, cleanup, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import EditProduct from './EditProduct';
import theme from '../../../styled/theme';
import { clickAtSubmitButton, createProductItem } from '../../../shared/testUtility/testUtility';
import { productConditions, productPhotoFieldValues } from '../../../shared/constants';
import * as actions from '../../../store/actions/indexActions';
import { mockFile } from '../../../shared/utility/utility';

const mockStore = configureMockStore([thunk]);

const defaultId = 'p1';
const defaultName = 'Wellingtons';
const defaultPrice = 5.2;
const defaultQuantity = 10;
const defaultCondition = productConditions.NEW;
const defaultDescription = 'Cool wellingtons';
const defaultPhoto = true;

const setUp = () => {
  const store = mockStore({
    product: {
      productDetails: createProductItem({
        id: defaultId,
        name: defaultName,
        price: defaultPrice,
        quantity: defaultQuantity,
        condition: defaultCondition,
        photo: defaultPhoto,
        description: defaultDescription,
      }),
    },
    ui: {
      isFormLoading: false,
      formError: '',
    },
  });
  store.dispatch = jest.fn();

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <EditProduct />
        </ThemeProvider>
      </Provider>,
    ),
    store,
  };
};

jest.mock('../../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../../store/actions/indexActions.js'),
  editProduct: (data, productId) => ({
    data,
    productId,
  }),
}));

afterEach(cleanup);

describe('<EditProduct />', () => {
  describe('check how renders', () => {
    it('should render everything correctly', () => {
      const { asFragment } = setUp();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('check form', () => {
    const defaultNewName = 'Boots';
    const defaultNewPrice = 25.25;
    const defaultNewQuantity = 25;
    const defaultNewDescription = 'Cool boots';
    const defaultNewPhoto = mockFile.create('boots.png', 1024, 'image/png');

    it('should have empty inputs and focus on email input by default', () => {
      setUp();
      const nameInput = screen.getByTestId('EditProduct-name');
      const priceInput = screen.getByTestId('EditProduct-price');
      const quantityInput = screen.getByTestId('EditProduct-quantity');
      const conditionNewInput = screen.getByTestId('EditProduct-condition-new');
      const conditionUsedInput = screen.getByTestId('EditProduct-condition-used');
      const conditionNotAppplicableInput = screen.getByTestId(
        'EditProduct-condition-not-applicable',
      );
      const descriptionInput = screen.getByTestId('EditProduct-description');

      expect(nameInput).toHaveFocus();
      expect(nameInput.value).toEqual(defaultName);
      expect(priceInput.value).toEqual(defaultPrice.toString());
      expect(quantityInput.value).toEqual(defaultQuantity.toString());
      expect(conditionNewInput).toBeChecked();
      expect(conditionUsedInput).not.toBeChecked();
      expect(conditionNotAppplicableInput).not.toBeChecked();
      expect(descriptionInput.value).toEqual(defaultDescription);
    });

    describe('should call', () => {
      it('should call editProduct() with updated all values after inputs submits and button click', async () => {
        const { store, container } = setUp();

        const nameInput = screen.getByTestId('EditProduct-name');
        const priceInput = screen.getByTestId('EditProduct-price');
        const quantityInput = screen.getByTestId('EditProduct-quantity');
        const conditionNewInput = screen.getByTestId('EditProduct-condition-new');
        const conditionUsedInput = screen.getByTestId('EditProduct-condition-used');
        const conditionNotAppplicableInput = screen.getByTestId(
          'EditProduct-condition-not-applicable',
        );
        const descriptionInput = screen.getByTestId('EditProduct-description');
        const uploadPhotoInput = screen.getByTestId('UploadPhoto-input');

        const dataToPass = {
          name: defaultNewName,
          price: defaultNewPrice,
          quantity: defaultNewQuantity,
          condition: productConditions.USED,
          description: defaultNewDescription,
          photo: defaultNewPhoto,
        };

        await waitFor(() => {
          fireEvent.change(nameInput, { target: { value: defaultNewName } });
        });
        await waitFor(() => {
          fireEvent.change(priceInput, { target: { value: defaultNewPrice } });
        });
        await waitFor(() => {
          fireEvent.change(quantityInput, { target: { value: defaultNewQuantity } });
        });
        await waitFor(() => {
          fireEvent.click(conditionUsedInput);
        });
        await waitFor(() => {
          fireEvent.change(uploadPhotoInput, { target: { files: [defaultNewPhoto] } });
        });
        await waitFor(() => {
          fireEvent.change(descriptionInput, { target: { value: defaultNewDescription } });
        });
        expect(nameInput.value).toEqual(defaultNewName);
        expect(priceInput.value).toEqual(defaultNewPrice.toString());
        expect(quantityInput.value).toEqual(defaultNewQuantity.toString());
        expect(conditionNewInput).not.toBeChecked();
        expect(conditionUsedInput).toBeChecked();
        expect(conditionNotAppplicableInput).not.toBeChecked();
        expect(descriptionInput.value).toEqual(defaultNewDescription);

        await clickAtSubmitButton(container);
        expect(store.dispatch).toHaveBeenNthCalledWith(
          1,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(nameInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          2,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(priceInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          3,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(quantityInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          4,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(conditionNewInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          5,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(conditionUsedInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          6,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(conditionNotAppplicableInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          7,
          actions.editProduct(dataToPass, defaultId),
        );

        await waitFor(() => {
          fireEvent.submit(descriptionInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          8,
          actions.editProduct(dataToPass, defaultId),
        );

        expect(store.dispatch).toHaveBeenCalledTimes(8);
      });

      describe('single values', () => {
        it('should call editProduct() with changed name only', async () => {
          const { store, container } = setUp();

          const nameInput = screen.getByTestId('EditProduct-name');
          const dataToPass = {
            name: defaultNewName,
          };

          await waitFor(() => {
            fireEvent.change(nameInput, { target: { value: defaultNewName } });
          });
          expect(nameInput.value).toEqual(defaultNewName);

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed price only', async () => {
          const { store, container } = setUp();

          const priceInput = screen.getByTestId('EditProduct-price');
          const dataToPass = {
            price: defaultNewPrice,
          };

          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: defaultNewPrice } });
          });
          expect(priceInput.value).toEqual(defaultNewPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed quantity only', async () => {
          const { store, container } = setUp();

          const quantityInput = screen.getByTestId('EditProduct-quantity');
          const dataToPass = {
            quantity: defaultNewQuantity,
          };

          await waitFor(() => {
            fireEvent.change(quantityInput, { target: { value: defaultNewQuantity } });
          });
          expect(quantityInput.value).toEqual(defaultNewQuantity.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed condition only (use labels to pick condition)', async () => {
          const { store, container } = setUp();

          const conditionNewInput = screen.getByTestId('EditProduct-condition-new');
          const conditionUsedInput = screen.getByTestId('EditProduct-condition-used');
          const conditionNotAppplicableInput = screen.getByTestId(
            'EditProduct-condition-not-applicable',
          );

          const dataToPass = {
            condition: productConditions.NOT_APPLICABLE,
          };

          await waitFor(() => {
            fireEvent.click(screen.getByText(/used/i));
          });
          expect(conditionNewInput).not.toBeChecked();
          expect(conditionUsedInput).toBeChecked();
          expect(conditionNotAppplicableInput).not.toBeChecked();

          await waitFor(() => {
            fireEvent.click(screen.getByText(/new/i));
          });
          expect(conditionNewInput).toBeChecked();
          expect(conditionUsedInput).not.toBeChecked();
          expect(conditionNotAppplicableInput).not.toBeChecked();

          await waitFor(() => {
            fireEvent.click(screen.getByText(/not applicable/i));
          });
          expect(conditionNewInput).not.toBeChecked();
          expect(conditionUsedInput).not.toBeChecked();
          expect(conditionNotAppplicableInput).toBeChecked();

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed description only', async () => {
          const { store, container } = setUp();

          const descriptionInput = screen.getByTestId('EditProduct-description');
          const dataToPass = {
            description: defaultNewDescription,
          };

          await waitFor(() => {
            fireEvent.change(descriptionInput, { target: { value: defaultNewDescription } });
          });
          expect(descriptionInput.value).toEqual(defaultNewDescription);

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed photo only (photo is deleted)', async () => {
          const { store, container } = setUp();
          const dataToPass = {
            photo: productPhotoFieldValues.DELETED,
          };

          await waitFor(() => {
            fireEvent.click(screen.getByRole('button', { name: /delete current/i }));
          });

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with changed photo only (photo is changed)', async () => {
          const { store, container } = setUp();

          const uploadPhotoInput = screen.getByTestId('UploadPhoto-input');
          const dataToPass = {
            photo: defaultNewPhoto,
          };

          await waitFor(() => {
            fireEvent.change(uploadPhotoInput, { target: { files: [defaultNewPhoto] } });
          });
          await waitFor(() => {});

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });
      });

      describe('check the lowest, highest and longest possible values', () => {
        it('should call editProduct() with max possible characters in name (price is changed)', async () => {
          const { store, container } = setUp();

          const nameInput = screen.getByTestId('EditProduct-name');

          let newName = '';
          for (let i = 1; i <= 150; i += 1) {
            newName += 'e';
          }
          const dataToPass = {
            name: newName,
          };

          await waitFor(() => {
            fireEvent.change(nameInput, { target: { value: newName } });
          });
          expect(nameInput.value).toEqual(newName);

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with lowest possible price (name is changed)', async () => {
          const { store, container } = setUp();

          const priceInput = screen.getByTestId('EditProduct-price');

          const newPrice = 0.01;
          const dataToPass = {
            price: newPrice,
          };

          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: newPrice } });
          });
          expect(priceInput.value).toEqual(newPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with highest possible price (name is changed)', async () => {
          const { store, container } = setUp();

          const priceInput = screen.getByTestId('EditProduct-price');

          const newPrice = 1000000;
          const dataToPass = {
            price: newPrice,
          };

          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: newPrice } });
          });
          expect(priceInput.value).toEqual(newPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with highest possible quantity (name is changed)', async () => {
          const { store, container } = setUp();

          const quantityInput = screen.getByTestId('EditProduct-quantity');

          const newQuantity = 100000;
          const dataToPass = {
            quantity: newQuantity,
          };

          await waitFor(() => {
            fireEvent.change(quantityInput, { target: { value: newQuantity } });
          });
          expect(quantityInput.value).toEqual(newQuantity.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });

        it('should call editProduct() with max possible characters in description (price is changed)', async () => {
          const { store, container } = setUp();

          const descriptionInput = screen.getByTestId('EditProduct-description');

          let newDescription = '';
          for (let i = 1; i <= 600; i += 1) {
            newDescription += 'e';
          }
          const dataToPass = {
            description: newDescription,
          };

          await waitFor(() => {
            fireEvent.change(descriptionInput, { target: { value: newDescription } });
          });
          expect(descriptionInput.value).toEqual(newDescription);

          await clickAtSubmitButton(container);

          expect(store.dispatch).toHaveBeenCalledWith(actions.editProduct(dataToPass, defaultId));
        });
      });
    });

    describe('should NOT call', () => {
      it('should NOT call editProduct() if all inputs have default values', async () => {
        const { store, container } = setUp();
        await clickAtSubmitButton(container);
        expect(store.dispatch).not.toHaveBeenCalled();
      });

      describe('values are too long, too high or too low', () => {
        it('should NOT call editProduct() if name is too long', async () => {
          const { store, container } = setUp();

          const nameInput = screen.getByTestId('EditProduct-name');

          let newName = '';
          for (let i = 1; i <= 151; i += 1) {
            newName += 'e';
          }

          await waitFor(() => {
            fireEvent.change(nameInput, { target: { value: newName } });
          });
          expect(nameInput.value).toEqual(newName);

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call editProduct() if price is too low', async () => {
          const { store, container } = setUp();

          const priceInput = screen.getByTestId('EditProduct-price');
          const newPrice = 0;

          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: newPrice } });
          });
          expect(priceInput.value).toEqual(newPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call editProduct() if price is too high', async () => {
          const { store, container } = setUp();

          const priceInput = screen.getByTestId('EditProduct-price');
          const newPrice = 1000000.01;

          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: newPrice } });
          });
          expect(priceInput.value).toEqual(newPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call editProduct() if quantity is too low', async () => {
          const { store, container } = setUp();

          const quantityInput = screen.getByTestId('EditProduct-quantity');
          const newQuantity = 0;

          await waitFor(() => {
            fireEvent.change(quantityInput, { target: { value: newQuantity } });
          });
          expect(quantityInput.value).toEqual(newQuantity.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call editProduct() if quantity is too high', async () => {
          const { store, container } = setUp();

          const quantityInput = screen.getByTestId('EditProduct-quantity');
          const newQuantity = 100001;

          await waitFor(() => {
            fireEvent.change(quantityInput, { target: { value: newQuantity } });
          });
          expect(quantityInput.value).toEqual(newQuantity.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call editProduct() if description is too long', async () => {
          const { store, container } = setUp();

          const descriptionInput = screen.getByTestId('EditProduct-description');

          let newDescription = '';
          for (let i = 1; i <= 601; i += 1) {
            newDescription += 'e';
          }

          await waitFor(() => {
            fireEvent.change(descriptionInput, { target: { value: newDescription } });
          });
          expect(descriptionInput.value).toEqual(newDescription);

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call addProduct() if photo is too big (price and name are changed)', async () => {
          const { store, container } = setUp();

          const nameInput = screen.getByTestId('EditProduct-name');
          const priceInput = screen.getByTestId('EditProduct-price');
          const uploadPhotoInput = screen.getByTestId('UploadPhoto-input');
          const newPhoto = mockFile.create('wellingtons.png', 6500000, 'image/png');

          await waitFor(() => {
            fireEvent.change(nameInput, { target: { value: defaultNewName } });
          });
          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: defaultNewPrice } });
          });
          await waitFor(() => {
            fireEvent.change(uploadPhotoInput, { target: { files: [newPhoto] } });
          });
          expect(nameInput.value).toEqual(defaultNewName);
          expect(priceInput.value).toEqual(defaultNewPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });

        it('should NOT call addProduct() if photo has incorrect extension (price and name are changed)', async () => {
          const { store, container } = setUp();

          const nameInput = screen.getByTestId('EditProduct-name');
          const priceInput = screen.getByTestId('EditProduct-price');
          const uploadPhotoInput = screen.getByTestId('UploadPhoto-input');
          const newPhoto = mockFile.create('wellingtons.png', 1024, 'image/svg+xml');

          await waitFor(() => {
            fireEvent.change(nameInput, { target: { value: defaultNewName } });
          });
          await waitFor(() => {
            fireEvent.change(priceInput, { target: { value: defaultNewPrice } });
          });
          await waitFor(() => {
            fireEvent.change(uploadPhotoInput, { target: { files: [newPhoto] } });
          });
          expect(nameInput.value).toEqual(defaultNewName);
          expect(priceInput.value).toEqual(defaultNewPrice.toString());

          await clickAtSubmitButton(container);

          expect(store.dispatch).not.toHaveBeenCalled();
        });
      });
    });
  });
});
