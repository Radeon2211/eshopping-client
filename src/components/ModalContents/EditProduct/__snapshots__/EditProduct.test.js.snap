// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProduct /> check how renders should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hYjfqi"
  >
    <h3
      class="sc-iBPRYJ jtekve"
    >
      Edit a product
    </h3>
    <form
      action="#"
      class="formik-form"
    >
      <div
        class="sc-fubCfw drZTsR"
        type="text"
      >
        <label
          class="sc-jrAGrp bkXExJ"
          for="name"
        >
          Name
        </label>
        <input
          autocomplete="off"
          class="sc-pFZIQ fjonmB"
          data-testid="EditProduct-name"
          id="name"
          name="name"
          placeholder="Name of the product (up to 150 characters)"
          type="text"
          value="Wellingtons"
        />
      </div>
      <div
        class="sc-crrsfI YLswv"
      >
        <div
          class="sc-fubCfw drZTsR"
          type="number"
        >
          <label
            class="sc-jrAGrp bkXExJ"
            for="price"
          >
            Price ($)
          </label>
          <input
            autocomplete="off"
            class="sc-pFZIQ fjonmB"
            data-testid="EditProduct-price"
            id="price"
            max="1000000"
            min="0.01"
            name="price"
            placeholder="Price of the product"
            step=".01"
            type="number"
            value="5.2"
          />
        </div>
        <div
          class="sc-fubCfw drZTsR"
          type="number"
        >
          <label
            class="sc-jrAGrp bkXExJ"
            for="quantity"
          >
            Quantity
          </label>
          <input
            class="sc-pFZIQ fjonmB"
            data-testid="EditProduct-quantity"
            id="quantity"
            max="1000000"
            min="1"
            name="quantity"
            placeholder="Quantity you want to sell"
            type="number"
            value="10"
          />
        </div>
      </div>
      <div
        class="sc-fubCfw drZTsR"
      >
        <label
          class="sc-jrAGrp bkXExJ"
        >
          Condition
        </label>
        <div
          class="sc-kEjbxe lbZuO"
        >
          <input
            checked=""
            class="sc-pFZIQ fjonmB"
            data-testid="EditProduct-condition-new"
            id="new"
            label="new"
            name="condition"
            type="radio"
            value="new"
          />
          <label
            for="new"
          >
            new
          </label>
        </div>
        <div
          class="sc-kEjbxe byHhTc"
        >
          <input
            class="sc-pFZIQ fjonmB"
            data-testid="EditProduct-condition-used"
            id="used"
            label="used"
            name="condition"
            type="radio"
            value="used"
          />
          <label
            for="used"
          >
            used
          </label>
        </div>
        <div
          class="sc-kEjbxe byHhTc"
        >
          <input
            class="sc-pFZIQ fjonmB"
            data-testid="EditProduct-condition-not-applicable"
            id="not_applicable"
            label="not applicable"
            name="condition"
            type="radio"
            value="not_applicable"
          />
          <label
            for="not_applicable"
          >
            not applicable
          </label>
        </div>
      </div>
      <div
        class="sc-fubCfw drZTsR"
      >
        <label
          class="sc-jrAGrp bkXExJ"
          for="description"
        >
          Description
        </label>
        <textarea
          class="textarea"
          data-testid="EditProduct-description"
          id="description"
          name="description"
          placeholder="Describe your product (up to 800 characters)"
        >
          Cool wellingtons
        </textarea>
      </div>
      <div
        class="sc-iqHYGH dfaNMv"
      >
        <div
          class="sc-gKsewC RnYls"
          direction="row"
          spacing="3"
          width="100%"
          wrap="nowrap"
        >
          <label
            class="label"
            for="photo"
          >
            <button
              class="sc-gsTCUz gQZzmy"
              color="blue"
              type="button"
            >
              Upload photo
            </button>
          </label>
          <button
            class="sc-gsTCUz gqkUyy"
            color="red"
            type="button"
          >
            Delete current
          </button>
        </div>
        <span
          class="sc-eCssSg gXddCR"
          color=""
          display="inline-block"
          overflow="visible"
          size="2"
          spacing="0px"
          transform=""
        >
          Photo is optional. Max size is 6 MB. Available extensions are JPG and PNG
        </span>
        <input
          class="input"
          data-testid="UploadPhoto-input"
          id="photo"
          name="photo"
          type="file"
        />
      </div>
      <div
        class="sc-gKsewC dLOeOj"
        direction="row"
        spacing="3"
        width="100%"
        wrap="nowrap"
      >
        <span
          class="sc-eCssSg gABWUT"
          color=""
          display="block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          <button
            class="sc-gsTCUz gNMffe"
            color="blue"
            type="button"
          >
            cancel
          </button>
        </span>
        <button
          class="sc-gsTCUz ghohWv"
          color="blue"
          data-testid="Form-submit-btn"
          disabled=""
          type="submit"
        >
          Edit
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
